### ISSUE TITLE
Proper fork to create PR against HA

### ISSUE THREAD

**Issue by danielperna84** [jazzaj/home-assistant-homematic#9]
To include our work into the official HA-package we need a proper HA-fork (repository) to which we commit the current code and add the needed tests.

How to do this is described here: https://home-assistant.io/developers/development_environment/

I propose we do this prior to continue working on other issues, because plainly copying the code from A to B won't include the commit-history.

In the long run, the Homematic-integration shouldn't require a dedicated repository besides the official HA-repo. Contributers usually will fork from there to commit changes, report issues etc.

@jazzaj Should I create the fork and take care of the tests, or do you want to handle the fork and I'll commit the tests to your repo?

**Comment by danielperna84** [comment]
Since I had the time today, I started by creating the fork myself. You can find it here: https://github.com/danielperna84/home-assistant/tree/Homematic

Unfortunately, tox currently doesn't pass.

I also started removing documentation from the individual files since the documentation belongs to the "real" documentation I mentioned in Issue #10 .

@jazzaj 
I had to modify the __ init __.py for the rollershutter. In your code you have multiple globals for the position. Is this intended? I stripped it out and just added the code below, using ATTR_CURRENT_POSITION instead of ATTR_POSITION.

**Comment by danielperna84** [comment]
First of all: I've readded the config-examples to the module docstrings so I can start the PR without relying on the proper documentation.

I have now fixed all lint-errors tox (`tox -e lint`) has complained about. There is 1 flake8-error remaining, but that can be ignored (error by design). And there are errors related to code we haven't touched, so I'll ignore that as well.

On my Windows 10 machine the runtime-tests (`tox -e py34` and `tox -e py35`) don't run at all because it's complaining about missing interpreters.
On my Debian machine the runtime-tests worked with py34, but I had to comment out the temperusb package in requirements_all.txt because of some ascii/unicode error while downloading that package. Again something not related to Homematic, and we don't have runtime-tests anyway.

Running `tox -e requirements` (Win 10 and Debian) results in an error where it **fails importing pyhomematic**. I assume it can't setup the pyhomematic package, maybe even fails at downloading it. This fits my observation, that pyhomematic didn't get installed automatically in a test-setup where I haven't installed pyhomematic manually.

One important note: pyhomematic has been updated to 0.1.2 because HA made some changes in how services (?) are stopped. Because of that shutting down HA didn't stop the pyhomematic server-thread. This is fixed with 0.1.2.

What I would need now is some testing by somebody else than me with a different setup and different devices.
@jazzaj When using my repo, do your automations still work? Especially the rollershutters, since I had to mess around with them. I re-added your changes since I now have understood what they where there for.
@Marioheld Your power-meter switch is now supported if you configure it as a switch. The power-meter functionality will be added later.
@pvizeli If you want it would be nice if you could test as well.

@all Feel free to try out the tox-tests for yourself. Maybe they work properly in your environments. We definitely have to figure out why pyhomematic doesn't get installed during the tox-tests.

**Comment by jazzaj** [comment]
@danielperna84: Highly appreciated you took care for the fork. 
You are right the ATTR_POSITION could be simplified as you did. But I would not advise it, as others could have build already scripts on the ATTR_CURRENT_POSITION (current_position) part. I would propose to have both to stay compatabile. 
It will take a couple of days until I can test. Do they work for your rollershutter?

**Comment by danielperna84** [comment]
Basic functionality of the my rollershutter works. I never tried seeking of specific positions, and since you have the experience with that, I wanted to leave that to you.

**Comment by Marioheld** [comment]
Can also confirm that my rollershutters work (2x HM-LC-Bl1PBU-FM controlled over Homegear -> HM-CFG-USB) also my switch works like a charm even updating when activating manually. Thanks a lot for the support here. I installed yesterday on a Raspberry Pi a brand new [image](https://www.homegear.eu/index.php/Downloads) and then installed HA with the new one script option and afterwards installed HA Homematic Branch from Github parallel as an development environment. So I think pyhomematic could be installed without a problem because I didn't installed it manually. Try the next days to start experimenting with tox. Also did you start with the documentation already?

**Comment by danielperna84** [comment]
Glad to hear everything is working for you.

Meanwhile the relevant tox tests all seem to pass.

Yes, I am already taking care of the documentation. Started it here: https://github.com/danielperna84/home-assistant.io

@pvizeli currently is making some restructuring in pyhomematic. It shouldn't affect configuration in HA, but I'll wait until he has finished his work before I continue with the documentation. Just in case.

**Comment by Marioheld** [comment]
Great to hear! Can't wait to get the homematic plugin in HA for default. Are there plans for integrating the existing homematic thermostat in this plugin? It would be easier for set up and also less confusing for new users.

**Comment by danielperna84** [comment]
Yes, the thermostat is already implemented. It actually should be working with the code at https://github.com/danielperna84/home-assistant. Just configure it something like this:

``` yaml
thermostat:
  platform: homematic
  address: "KEQ1234567"
  name: "Kitchen"
```

The repository above is the one from where we'll do the pull request - which is the "solution" for this issue. ;-)

**Comment by Marioheld** [comment]
Sorry didn't see that. :D I have one thermostat laying around l if it need some testing.

**Comment by danielperna84** [comment]
I have a thermostat myself, so I can already confirm it is working. What wh haven't tested are the wall-mounted thermostats. But @pvizeli has one of those, so he can report what's working and what is not. He currently is making some change to the pyhomematic library, which will enable smoother HA-integration.

**Comment by danielperna84** [comment]
To keep everyone updated:
@pvizeli has made some major improvements to pyhomematic. This in turn though requirese some modification of the current HA-integration. He will take care of this and probably present his results during this week.
Until then you should NOT try update pyhomematic to a current release, as it will break Homematic within HA.

**Comment by danielperna84** [comment]
We have created the PR against HA. Instead of this repo @pvizeli created his own one, because this one meanwhile was pretty very from HAs dev-branch. It made more sense that pvizeli created a clean fork with our code plugged in. His fork can be found here: https://github.com/pvizeli/home-assistant/tree/Homematic
With his repo A LOT of devices work, and even auto-discovery works pretty well. But we still consider it experimental at the moment. It is able to detect and configure all devices we currently support. And it does so only, if a device hasn't been configured manually. So a nice process would be to start off with no devices configured, see which devices are available as whatever type it got auto-configured, and then, if wanted, reproduce that cofiguration manually to give the entity a useful name.
I'm working on the documentation now. Some devices might be complex to configure if you don't know how to get all features out of them.

**Comment by danielperna84** [comment]
If you need some guidance on how to setup or what's possible, look for the files with homematic in their name here: https://github.com/danielperna84/home-assistant.io/tree/next/source/_components
The documentation for the component itself seems done to me. To do are the platforms. I'm working on that currently. The one for binary sensors is done already. The ones for rollershutters, switches and thermostats are prepared with configuration examples but no explanation.

**Comment by danielperna84** [comment]
Documentation is now complete. You can find it here:
https://github.com/danielperna84/home-assistant.io/tree/next/source/_components

What's new since my last post:
There are two new config parameters for the component: resolvenames and delay.
When you have a lot of devices HA queries the CCU while booting A LOT, which resulsts in "Servicemeldungen" filling up. This may be a bad thing. For this reason we have decided to set a per device delay of 0.5 seconds to fetch the current state. For me this results in now more messages in the Servicemeldungen. If you only have a hand full of devices you can lower this value to something like 0.1, or raise it to 1 if you are rich and have so many devices even 0.5 messes with your CCU. 

The second option, resolvenames, tries to fetch the device name from either the metadata you can set when using HM-CFG-LAN, or from the CCU if you have the XML-API installed. Using the Metadata from the HM-CFG-LAN works very well. I couldn't test the XML-API variant.
Because the CCU internally has metadata as well, we always try that first. Then we try to connect to the XML-API. If that is successful, the previously fetched names will be overwritten. This feature is especially great for people who just want to try out Homematic within HA.

**Comment by danielperna84** [comment]
Again an update:
Homematic has been merged into HAs dev-branch, and will be released together with the upcoming version 0.23!

We've made an important design-decision: we only support auto-detection of devices. Allowing both manual configuration and autodetection has proven to be unreliable. And since manually configuring devices can be tricky, we chose to always auto-detect. Using the [Customizing](https://home-assistant.io/getting-started/customizing-devices/) feature of HA, it is very easy to rename the automatically created entities to something you like. Additionally you can hide entities, which is useful for binary_sensor like the buttons of a remote control.

So essentially the initial Homematic-configuration can be done with 4 lines of yaml and will be fully functional. We think this is way more user-friendly than having to configure everything manually, and especially users new to HA (the ones stuck on FHEM, OpenHAB etc.) will be able to get started very quickly.

Now we're just waiting for 0.23 to be released. :)

**Comment by danielperna84** [comment]
HA 0.23 has been released today. :tada:
